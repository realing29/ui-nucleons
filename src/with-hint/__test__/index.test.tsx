import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import { WithHint, useTempHint } from '..';
import { Hint } from '../../hint';

describe('<WithHint />', () => {
  it('should renders correctly', () => {
    const wrapper = mount(
      <WithHint hint='Hello, world'>
        {(ref, toggle) => (
          <div
            ref={ref as any}
            data-testid='test-opener'
            onMouseEnter={() => toggle(true)}
            onMouseLeave={() => toggle(false)}
          >Opener</div>
        )}
      </WithHint>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Hint)).toHaveLength(0);

    act(() => {
      wrapper.find('[data-testid="test-opener"]').simulate('mouseenter');
    });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Hint)).toHaveLength(1);

    act(() => {
      wrapper.find('[data-testid="test-opener"]').simulate('mouseleave');
    });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Hint)).toHaveLength(0);
  });
});

describe('useTempHint', () => {
  const TestComponent = () => {
    const [shown, toggle] = useTempHint();

    return (
      <div data-testid='test-opener' onClick={() => toggle(true)}>
        {shown ? 'shown' : 'hidden'}
      </div>
    );
  };

  jest.useFakeTimers();

  it('should works', () => {
    const wrapper = mount(
      <TestComponent />
    );

    expect(wrapper.text()).toBe('hidden');

    act(() => {
      wrapper.find('[data-testid="test-opener"]').simulate('click');
    });
    wrapper.update();

    expect(wrapper.text()).toBe('shown');

    act(() => {
      jest.advanceTimersByTime(2500);
    });
    wrapper.update();

    expect(wrapper.text()).toBe('hidden');
  });

  it('should recreate timer', () => {
    const wrapper = mount(
      <TestComponent />
    );

    expect(wrapper.text()).toBe('hidden');

    // try show
    act(() => {
      wrapper.find('[data-testid="test-opener"]').simulate('click');
    });
    wrapper.update();

    expect(wrapper.text()).toBe('shown');

    // wait
    act(() => {
      jest.advanceTimersByTime(500);
    });
    wrapper.update();

    expect(wrapper.text()).toBe('shown');

    // try show again
    act(() => {
      wrapper.find('[data-testid="test-opener"]').simulate('click');
    });
    wrapper.update();

    expect(wrapper.text()).toBe('shown');

    // wait for first timer
    act(() => {
      jest.advanceTimersByTime(1300);
    });
    wrapper.update();

    expect(wrapper.text()).toBe('shown');

    // wait
    act(() => {
      jest.advanceTimersByTime(2200);
    });
    wrapper.update();

    expect(wrapper.text()).toBe('hidden');
  });
});
