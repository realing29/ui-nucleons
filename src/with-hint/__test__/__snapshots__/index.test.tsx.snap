// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WithHint /> should renders correctly 1`] = `
<WithHint
  hint="Hello, world"
>
  <div
    data-testid="test-opener"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Opener
  </div>
</WithHint>
`;

exports[`<WithHint /> should renders correctly 2`] = `
<WithHint
  hint="Hello, world"
>
  <div
    data-testid="test-opener"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Opener
  </div>
  <Portal>
    <Portal
      containerInfo={
        <div>
          <div
            class="root hint top"
            data-testid="hint"
            style="z-index: 1; left: 0px; top: -4px;"
          >
            <div
              class="body shadow-z4"
            >
              Hello, world
            </div>
          </div>
        </div>
      }
    >
      <PositioningHint
        direction="top"
        openerRef={
          {
            "current": <div
              data-testid="test-opener"
            >
              Opener
            </div>,
          }
        }
      >
        <ForwardRef
          className="hint hidden"
          direction="top"
          style={
            {
              "zIndex": 1,
            }
          }
        >
          <div
            className="root hint hidden top"
            data-testid="hint"
            style={
              {
                "zIndex": 1,
              }
            }
          >
            <div
              className="body shadow-z4"
            >
              Hello, world
            </div>
          </div>
        </ForwardRef>
      </PositioningHint>
    </Portal>
  </Portal>
</WithHint>
`;

exports[`<WithHint /> should renders correctly 3`] = `
<WithHint
  hint="Hello, world"
>
  <div
    data-testid="test-opener"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Opener
  </div>
</WithHint>
`;
